utils.py

```diff
@@ -728,6 +729,109 @@ def get_hashable_value(value):
+        if ca_spec.name not in customization_args:
+            customization_args[ca_spec.name] = {
+                'value': ca_spec.default_value
+            }
+    return customization_args
+
+
+def validate_customization_args_and_values(
+        item_name, item_type, customization_args,
+        ca_specs_to_validate_against):
+    """Validates the given `customization_args` dict against the specs set out
+    in 'ca_specs_to_validate_against'. 'item_name' and 'item_type' are used to
+    populate any error messages that arise during validation.
+    Note that this may modify the given customization_args dict, if it has
+    extra or missing keys. It also normalizes any HTML in the
+    customization_args dict.
+
+    Args:
+        item_name: str. This is always 'interaction'.
+        item_type: str. The item_type is the ID of the interaction.
+        customization_args: dict. The customization dict. The keys are names of
+            customization_args and the values are dicts with a
+            single key, 'value', whose corresponding value is the value of
+            the customization arg.
+        ca_specs_to_validate_against: list(dict). List of spec dictionaries. Is
+            used to check if some keys are missing in customization_args. Dicts
+            have the following structure:
+                - name: str. The customization variable name.
+                - description: str. The customization variable description.
+                - default_value: *. The default value of the customization
+                    variable.
+
+    Raises:
+        ValidationError: The given 'customization_args' is not valid.
+    """
+    ca_spec_names = [
+        ca_spec.name for ca_spec in ca_specs_to_validate_against]
+
+    if not isinstance(customization_args, dict):
+        raise ValidationError(
+            'Expected customization args to be a dict, received %s'
+            % customization_args)
+
+    # Validate and clean up the customization args.
+
+    # Populate missing keys with the default values.
+    customization_args = get_full_customization_args(
+        customization_args, ca_specs_to_validate_against)
+
+    # Remove extra keys.
+    extra_args = []
+    for arg_name in customization_args.keys():
+        if not isinstance(arg_name, basestring):
+            raise ValidationError(
+                'Invalid customization arg name: %s' % arg_name)
+        if arg_name not in ca_spec_names:
+            extra_args.append(arg_name)
+            logging.warning(
+                '%s %s does not support customization arg %s.'
+                % (item_name.capitalize(), item_type, arg_name))
+    for extra_arg in extra_args:
+        del customization_args[extra_arg]
+
+    # Importing this at the top of the file causes a circular dependency
+    # as follows:
+    # utils -> schema_utils -> html_cleaner -> rte_component_registry -> utils.
+    import schema_utils # pylint: disable=relative-import
```

Comment:
Hmm, this will definitely bite a developer one day, so let's break the cycle sooner rather than later.

From looking through each file, the dependencies are:

- `utils` ->
   - `schema_utils.normalize_against_schema`
- `schema_utils` ->
   - `html_cleaner.clean`
   - `html_cleaner.filter_a`
- `html_cleaner` ->
   - `rte_component_registry.Registry.get_tag_list_with_attrs`
- `rte_component_registry` ->
   - `utils.camelcase_to_hyphenated`

I find the cleanest solution to be placing this into `schema_utils` instead of `utils`. Conceptually this makes sense since `customization_args` is an object whose value is defined by a 'schema', so placing it in `schema_utils` seems logical enough for me.

I also think that `utils` is "too general" anyway, there's nowhere else in the code base I can imagine that'd need this function besides one that is already using `schema_util`.

Reply:
Got it, Done!